src/config/include/config/config.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+// TODO: proper header
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <utility>
+
+#include <boost/log/trivial.hpp>
+
+#include <yaml-cpp/yaml.h>
+
+namespace genny {
+
+// We don't actually care about YAML::convert<T>
+// because we never go from T to YAML::Node just the
+// other way around.
+//
+// This template *may* be specialized by the user if
+// they wish to do special post-processing (e.g. if
+// the O type is a ptr-type and want to do dynamic_cast)
+template <typename O>
+struct NodeConvert {
+    static O convert(YAML::Node node) {
+        return node.as<O>();
+    }
+};
+
+template <typename C>
+class NodeT {
+    YAML::Node _yaml;
+    C* _parent;
+
+    // TODO: keep track of the key we came from so we can report it in error-messages
+    // using KeyType = std::optional<std::variant<std::string, int>>;
+    // KeyType _key;
+
+
+    template <typename K>
+    std::optional<YAML::Node> yamlGet(const K& key) {
+        YAML::Node found = _yaml[key];
+        if (found) {
+            return std::make_optional(found);
+        } else {
+            if (!_parent) {
+                return std::nullopt;
+            }
+            return _parent->yamlGet(key);
+        }
+    }
+
+    template <typename K>
+    NodeT get(const K& key) {  // TODO: const version
```

Comment:
I got it to compile by having the contexts inherit NodeT's constructor and making two changes mentioned below.

The first one is to remove the ".." key because the parent isn't always define. It seems like the ".." is only going to be used in code (and not static YAML files)? So maybe it could be replaced with ->parent()?